Newsletters: add 100 (2020-06-03) (#410)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2020-06-03-newsletter.md

---
title: 'Bitcoin Optech Newsletter #100'
permalink: /en/newsletters/2020/06/03/
name: 2020-06-03-newsletter
slug: 2020-06-03-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter summarizes a proposed design for a coinswap
implementation, describes new middleware for allowing lightweight

> wallets to get information directly from a user's own node, and links to

Reviewer: perhaps: s/get/fetch|retrieve/

two transaction size calculators.  Also included are our regular
sections with descriptions of several recently transcribed talks, new
releases and release candidates, and notable changes to popular Bitcoin
infrastructure software.  A special final section celebrates the
publication of Newsletter #100.

## Action items

*None this week.*

## News

- **Design for a coinswap implementation:** Chris Belcher
  [posted][belcher coinswap] a design for a full-featured coinswap
  implementation.  Coinswap is a protocol that allows two users to

>   create a pair of transaction that look like regular payments but which

Reviewer: s/transaction/transactions/

  actually swap their coins with each other.  This improves the privacy
  of not just the coinswap users but all Bitcoin users, as anything that
  looks like a payment could have instead been a coinswap.

    Belcher's post summarizes the history of the coinswap idea, suggests
    ways the multisig conditions needed for coinswap could be disguised
    as more common transaction types, proposes using a market for
    liquidity (like JoinMarket already does), describes splitting and
    routing techniques to reduce privacy losses from amount correlation
    or spying participants, mentions alternative coinswap protocols such
    as succinct atomic swaps (see [Newsletter #98][news98 sas]), suggests
    combining coinswap with [payjoin][topic payjoin], and discusses some
    of the backend requirements for the system.  Additionally,

>     comparisons are made to other privacy techniques such as using LN,

Reviewer: nit: perhaps active voice > passive voice here

Reviewer: A lot to unpack; maybe worth testing a separation of this paragraph into three:

(1) the post begins by...
(2) the post then goes on to compare...
(3) near the end...

Author: I think it's ok as one paragraph; the goal of the paragraph isn't to teach people anything but simply to provide a large number reasons that they should go read the original email.  If it feels like there's a lot to unpack, then I think that means it's mission accomplished.  :-)

    [coinjoin][topic coinjoin], payjoin, and [payswap][zmn payswap].
    Near the end of his email, Belcher summarizes the requirements for
    the system he describes:

>     > * [Two-party ECDSA]

Author: Eh, I think we don't actually need the list at all, so I'll just remove it.

    > * Liquidity market
    > * Routed CoinSwaps
    > * Multi-transaction CoinSwaps
    > * Breaking change output heuristics
    > * Fidelity bonds

>     > * PayJoin with CoinSwap

Reviewer: style guide : coinswap? (throughout, not just here)

Author: The PR already adds "coinswap" to the style guide as a common noun.  However, I left these in pascal case since they're direct quotes from Belcher and I like to respect author choices unless they're confusing or otherwise problematic.

>     > * Federated message boards protected from spam with fidelity bonds

Reviewer: perhaps s/with/by/

>     Belcher has a history of creating and maintaining privacy enhancing

Reviewer: privacy-enhancing (with hyphen)?

    notes that future versions of BWT may support [output script
    descriptors][topic descriptors], allowing wallets to produce and
    consume standardized descriptions of their script templates.

- **Transaction size calculators:** Jameson Lopp [posted][lopp size] to
  the Bitcoin-Dev mailing list with links to a [transaction size
  calculator][lopp calc] he'd developed as well as a similar [calculator
  developed by Optech][optech calc].  Neither tool claims to be complete
  or bug-free, but both should be useful to developers who want to
  quickly compare the sizes of different types of transactions.

## Recently transcribed talks and conversations

*[Bitcoin Transcripts][] is the home for transcripts of technical
Bitcoin presentations and discussions. In this monthly feature, we
highlight a selection of the transcripts from the previous month.*

- **LN backups:** Christian Decker presented at Potzblitz on the latest
  state of LN backups. He discussed the approaches of other
  implementations such as Eclair and LND before explaining why
  C-Lightning is using a synchronous database log plugin. Later, he
  described why LN backups are more complex than onchain backups, the

>   prospects of getting [SIGHASH_NOINPUT][topic sighash_noinput] or

Reviewer: maybe prospects of s/getting XYZ into Bitcoin/having XYZ merged into Bitcoin/

Author: @michaelfolkson actually used "merge [...] into Bitcoin Core" in his original text, but I changed it to the current phrasing because I don't like the implication that the focus should be on merging soft fork changes into Bitcoin Core.  IMO, the focus should be on convincing people that the changes should be made to the Bitcoin Protocol, for which Bitcoin Core is the current reference implementation.

I'm guessing your main complaint here, though, is about the word "getting", so I'm rephrasing it to "[...] of adding [...] to Bitcoin".  Is that acceptable?

Reviewer: That reasoning makes a lot of sense. Change looks good!

  `SIGHASH_ANYPREVOUT` to Bitcoin to enable [eltoo-based][topic eltoo]
  payment channels, and the current modularity of the LN protocol.
  ([transcript][decker xs], [video][decker vid])

- **Payjoin/P2EP:** Adam Gibson led a discussion at London BitDevs about

>   [payjoin][topic payjoin], including the recent implementation of it by

Reviewer 2: "the recent implementation of it by BTCPay Server" feels a bit convoluted. I think you can just drop this entirely from the sentence, since almost exactly the same words are repeated below.

>   BTCPay Server.  Payjoin allows the sender and receiver of a

Reviewer 2: I think this sentence could be shortened and made clearer by removing the unnecessary "to coordinate in order" and the repetitive "allows". Something like: "Payjoin allows a transaction's sender and receiver to both contribute inputs the transaction. This breaks the common wallet ownership assumption and subset sum analysis, and improves the sender's and receiver's privacy."

>   BTCPay Server.  Payjoin allows both the sender and receiver of a

Reviewer: I think "both" can be removed here, as it is understood and also present a few words later.

perhaps s/coordinate/cooperate/

  inputs to the transaction, breaking the [common wallet ownership
  assumption][] and subset sum analysis. Gibson went through the history
  of the concept and described the existing implementations of payjoin
  in JoinMarket and Samourai before examining details of the recent
  BTCPay Server implementation. He ended by outlining several different
  ways a wallet can be fingerprinted, such as the number of signatures
  required, what timelocks are used, and whether the opt-in
  Replace-By-Fee ([RBF][topic rbf]) flag is set.  ([transcript][gibson
  xs], [video][gibson vid])

- **LSAT---Your Ticket Aboard The Lightning Native Web:** Oliver Gugger
  presented on the Lightning Service Authentication Token (LSAT) at
  Reckless VR in virtual reality. LSAT is a proposed protocol
  specification combining HTTP, macaroons, and Lightning. It's designed
  to fulfill the purpose of the HTTP 402 Payment Required response code.
  Gugger described the authentication flow and the role of macaroons as
  pseudonymous user authentication. The question and answer session
  focused on use cases and the benefits of using LSAT, such as enhanced

>   user privacy and improved sign up experience.

Reviewer: s/sign up/sign-up/?

  ([transcript](https://diyhpl.us/wiki/transcripts/vr-bitcoin/2020-05-16-oliver-gugger-lsat/),
  [video](https://www.youtube.com/watch?v=IW08RJUpzw0))

- **Sydney meetup discussion:** A number of Bitcoin and LN developers
  joined this Sydney meetup to discuss topics including: the scalability
  issues of onboarding millions of LN clients, Rust code integration
  into Bitcoin Core, dual funding in C-Lightning, and modern soft fork
  activation. The history of Linux kernel development and segwit

>   activation were explored for insights into when or whether the Rust

Reviewer: If this is the history of both (it reads as though it is), then s/were/was/

  language should be introduced to Bitcoin Core and the exact mechanism
  that should be used to activate proposed soft forks such as

>   [taproot][topic taproot] in future. The transcript was anonymized to

Reviewer: s/in future/in the future/

  encourage open discussion.  ([transcript][sydney xs])

> - **Revault—A Multiparty Vault Architecture:** Kevin Loaec and Antoine

Reviewer: This may be correct and only just me, but I initially wondered what Revault—A was.

Maybe s/Revault—A Multiparty/Revault — A Multiparty/

>   Poinsot presented on their vault design *Revault* at London Bitcoin

Reviewer: Here and line 128 above, I wonder if s/presented on/presented/ would not be better.

Author: Nice improvement, I always like fewer words, thanks!

  Devs. They outlined specific details such as its reliance on
  co-signing servers and how it compares to some other vault designs
  that require key deletion, anticipating spending amounts, or both.
  Their presentation was preceded the week before by a wider discussion
  on vaults, [covenants][topic covenants], and `OP_CHECKTEMPLATEVERIFY`
  (`OP_CTV`). This included the merits of the different use cases of
  `OP_CTV` and a possible path to it being soft forked into Bitcoin.

>   Additional discussion focused on the current state of the the mempool

Reviewer: s/the the/the/

  policy in Bitcoin Core and how it creates challenges such as
  [transaction pinning][topic transaction pinning] that affect the
  security of both vault designs and the LN protocol.  ([Meetup
  transcript][london xs], [Presentation transcript][revault xs], [Meetup
  video][london vid], [Presentation video][revault vid])

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [Bitcoin Core 0.20.0rc2][bitcoin core 0.20.0] is the most recent
  release candidate for the next major version of Bitcoin Core.

- [LND 0.10.1-beta.rc3][] is the latest release candidate for the next
  maintenance release of LND.

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1 repo],
[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
BOLTs][bolts repo].*

*Note: the commits to Bitcoin Core mentioned below apply to its master

> development branch and so those changes will likely not be released

Reviewer: I think "so those changes" can be omitted (or replaced by ", and unless backported,")

Author: Backporting is one concern, but my main concern is promising that X merged feature will be in release Y.  There's always the chance that something will be reverted before then.  I agree that this warning is probably more verbose than necessary given that all reasonable developers understand that things can change between a development merge and a release---but a few years ago I polled some Bitcoin developers for some deadlines on segwit stuff, used that to make a table of planned release dates, and then saw /r/btc republish that table over and over again in attacks on the project when those dates weren't met.  Several core devs have told me they don't blame me for what happened, but ever since I've been *very* careful about writing anything that sounds like a promise to release specific features, much less promises to release at a certain point in time.

until version 0.21, about six months after the release of the upcoming
version 0.20.*

- [Bitcoin Core #19010][] net processing: Add support for getcfheaders FIXME:jonatack

> - [Bitcoin Core #16030][] changes how long Bitcoin Core waits until

Reviewer 2: s/16030/16939

Author: Ooh, :facepalm:, thanks!

  minutes before querying.  This improves the chance that a restarted
  node will entirely use P2P address discovery without relying on the
  centralized DNS seeds.

- [LND #4228][] walletrpc: add LabelTransaction endpoint to retrospectively label txns FIXME:dongcarl

## On the occasion of Optech Newsletter #100

> "I am somewhat surprised that nobody has taken to [writing] weekly
> summaries of research and development activity. Summarizing recent
> work is a valuable task that others can engage in just by reading the
> mailing list and aggregating multiple thoughts together."
>
> {:.right}
> ---Bryan Bishop, [19 August 2015][bishop summaries]

{% comment %} <!--
  $ find _posts/en/newsletters/ _includes/specials/bech32/  -type f | xargs cat | wc -w
  155217

  443 pages = 155217 words / 350 words per printed page
-->{% endcomment %}

Almost five years after Bishop made the above comment, we remain
convinced that writing weekly summaries of research and development
activity is a task that's valuable to both the open source Bitcoin
development community and to the many businesses that depend upon the
community's work.  But in the two years we've been producing this

> newsleter, we've also discovered that summarizing isn't quite as quick

Reviewer: newsletter

  required, what timelocks are used, and whether the opt-in
  Replace-By-Fee ([RBF][topic rbf]) flag is set.  ([transcript][gibson
  xs], [video][gibson vid])

- **LSAT---your ticket aboard the lightning native web:** Oliver Gugger
  presented the Lightning Service Authentication Token (LSAT) at
  Reckless VR in virtual reality. LSAT is a proposed protocol
  specification combining HTTP, macaroons, and Lightning. It's designed
  to fulfill the purpose of the HTTP 402 Payment Required response code.
  Gugger described the authentication flow and the role of macaroons as
  pseudonymous user authentication. The question and answer session
  focused on use cases and the benefits of using LSAT, such as enhanced
  user privacy and improved sign-up experience.
  ([transcript](https://diyhpl.us/wiki/transcripts/vr-bitcoin/2020-05-16-oliver-gugger-lsat/),
  [video](https://www.youtube.com/watch?v=IW08RJUpzw0))

- **Sydney meetup discussion:** A number of Bitcoin and LN developers
  joined this Sydney meetup to discuss topics including: the scalability
  issues of onboarding millions of LN clients, Rust code integration
  into Bitcoin Core, dual funding in C-Lightning, and modern soft fork

>   activation. The history of Linux kernel development and segwit

Reviewer 3: Modern isn't really the right word here. It's more like "possible soft fork activation methods".

  activation were each explored for insights into when or whether the Rust
  language should be introduced to Bitcoin Core and the exact mechanism
  that should be used to activate proposed soft forks such as
  [taproot][topic taproot] in the future. The transcript was anonymized to
  encourage open discussion.  ([transcript][sydney xs])

- **The Revault multiparty vault architecture:** Kevin Loaec and Antoine
  Poinsot presented their vault design *Revault* at London Bitcoin
  Devs. They outlined specific details such as its reliance on
  co-signing servers and how it compares to some other vault designs
  that require key deletion, anticipating spending amounts, or both.

>   Their presentation was preceded the week before by a wider discussion

Reviewer 3: ```suggestion
  Their presentation was preceded the week before by a broader discussion
```

    `getcfheaders`, and `getcfilters` requests with the corresponding `cfcheckpt`,
    `cfheaders`, and `cfilters` responses. The node does not yet advertise support
    for BIP157 with `NODE_COMPACT_FILTERS` in its version message. The final step,
    [Bitcoin Core #19070][], is under review at the time of this writing and
    would enable nodes to signal the ability to serve compact block filters.
    The feature is disabled by default and can be enabled with the
    `-peerblockfilters` configuration parameter.

- [Bitcoin Core #16939][] changes how long Bitcoin Core waits until
  it queries DNS seeds for the IP addresses of potential peers.
  Previously, if the node had peer IP addresses in its database, it
  would try opening several connections and wait 11 seconds for
  successful connections before requesting new addresses.  Now, if it
  has more than 1,000 IP addresses in its database---which is common for
  nodes that have been online more than a few hours---it'll wait up to 5
  minutes before querying.  This improves the chance that a restarted
  node will entirely use P2P address discovery without relying on the
  centralized DNS seeds.

> - [LND #4228][] adds a new wallet command `labeltx` for labelling past onchain

Reviewer 3: Grepped for our use of labeling and labelling in the code and found this to be the first instance of the non-American variant and three instances in the RBF in the Wild post.

```suggestion
- [LND #4228][] adds a new wallet command `labeltx` for labeling past onchain
```

>   transactions. This is a continuation of the work done in [LND #4213][] which

Reviewer 3: ```suggestion
  transactions. This is a continuation of the work done in [LND #4213][], which
```

  allowed setting a label when sending a payment.  Labels are personal
  wallet metadata that help the user remember who they paid and what
  they bought; the labels aren't part of the onchain transactions and
  aren't shared with any other user.

## On the occasion of Optech Newsletter #100

> "I am somewhat surprised that nobody has taken to [writing] weekly
> summaries of research and development activity. Summarizing recent
> work is a valuable task that others can engage in just by reading the
> mailing list and aggregating multiple thoughts together."
>
> {:.right}
> ---Bryan Bishop, [19 August 2015][bishop summaries]

{% comment %} <!--
  $ find _posts/en/newsletters/ _includes/specials/bech32/  -type f | xargs cat | wc -w
  155217

  443 pages = 155217 words / 350 words per printed page
-->{% endcomment %}

Almost five years after Bishop made the above comment, we remain
convinced that writing weekly summaries of research and development
activity is a task that's valuable to both the open source Bitcoin
development community and to the many businesses that depend upon the

> community's work.  But in the two years we've been producing this

Reviewer: s/But in/In/

(this works thanks to "also" in the sentence)

newsletter, we've also discovered that summarizing isn't quite as quick
and simple as we initially expected it to be.  Accordingly, we'd like to
take this chance to thank the people who make this newsletter possible

> by generously contributing a significant amount of their valuable time

Reviewer: s/time week after week/time, week after week/

> week after week: [Adam Jonas][], [Carl Dong][], [David A. Harding][],

Reviewer 3: I have to say I'm uncomfortable being mentioned first on this line with @harding buried in the middle (I get it's alphabetical). I don't believe that credit should be equally shared. Can emphases or an additional sentence be added to make it clear who does most of the work to make this newsletter happen?

Author: I'm sorry this makes you feel uncomfortable.  :-(  However, an alphabetical list of contributors in the standard for lists of credits in most free software projects even though all of those projects have the same situation of different people contributing different amounts of time to the project.  I think we're better off following that simple and pragmatic standard than we are trying to fairly apportion credit among ourselves.

Reviewer 3: @harding Just feel that you deserved to be highlighted since you put in the most work.

[John Newbery][], [Jon Atack][], [Mike Schmidt][], and [Steve Lee][].

We additionally thank the experienced Bitcoin and LN contributors who

> kindly provided us with special help on certain complex topics or who

Reviewer 3: 2 specials in the sentence.

```suggestion
kindly provided us with help on certain complex topics or who
```

contributed field reports and other special content to the newsletter
over the past two years.

> Publishing a high-quality weekly newsletter and working to fulfill other

Reviewer 3: ```suggestion
Publishing a high-quality, weekly newsletter and working to fulfill other
```

Author: This seems better to me without the comma, but I admit I'm not sure enough about the grammar rules to know which way is technically correct.

aspects of Optech's [mission][about page] wouldn't be possible without
the financial support of our member companies.  We thank them for their
continuing commitment to improving communication between Bitcoin users,
developers, and businesses.

{:.center style="font-size: 1.5em"}
![Thank you to our members!](/img/posts/2020-06-members.png)

We also remain eternally thankful to our founding sponsors [Wences
Casares][], [John Pfeffer][], and [Alex Morcos][] <!-- same order as on
About page --> as well as to organizations such as [Chaincode Labs][]

> and [SquareCrypto][] who both allow and encourage their staff to use

Reviewer 3: ```suggestion
and [Square Crypto][] who both allow and encourage their staff to use
```

> and [SquareCrypto][] who both allow and encourage their staff to use

Reviewer: suggest omitting "both", it's ambiguous whether it refers to both orgs or both "allow and encourage", and both of the meanings are understood without it

[news98 sas]: /en/newsletters/2020/05/20/#two-transaction-cross-chain-atomic-swap-or-same-chain-coinswap
[belcher coinswap]: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-May/017898.html
[zmn payswap]: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-January/017595.html
[joinmarket]: https://github.com/JoinMarket-Org/joinmarket-clientserver
[eps]: https://github.com/chris-belcher/electrum-personal-server
[lopp size]: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-May/017905.html
[lopp calc]: https://jlopp.github.io/bitcoin-transaction-size-calculator/
[optech calc]: https://bitcoinops.org/en/tools/calc-size
[about page]: /about/
[ivgi bwt]: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-May/017906.html
[bwt api]: https://github.com/shesek/bwt#http-api
[bwt plugin]: https://github.com/shesek/bwt#electrum-plugin
[edge]: https://edge.app/
[blue wallet]: https://bluewallet.io/
[eclair mobile]: https://github.com/ACINQ/eclair-mobile
[phoenix]: https://phoenix.acinq.co/
[Adam Jonas]: https://github.com/adamjonas
[Carl Dong]: https://github.com/dongcarl

> [David A.  Harding]: https://github.com/harding

Reviewer: extra space between A. and Harding, but the link is fine

  node will entirely use P2P address discovery without relying on the
  centralized DNS seeds.

- [LND #4228][] adds a new wallet command `labeltx` for labelling onchain
  transactions. This is a continuation of the work done in [LND #4213][] which

>   allowed setting a label option at send-initiation time.

Reviewer 4: Perhaps remove the "option"? What do people think?

Author: +1 for removing "option"

Reviewer 4: Done!
